README.md
# Textpattern support forum

[![Greenkeeper badge](https://badges.greenkeeper.io/textpattern/textpattern-forum.svg)](https://greenkeeper.io/)
[![Build Status](https://img.shields.io/travis/textpattern/textpattern-forum.svg)](https://travis-ci.org/textpattern/textpattern-forum)

This repository contains the source code for the [Textpattern support forum](https://forum.textpattern.io/).

## Supported web browsers

* Internet Explorer 11.
* Chrome, Edge, Firefox, Safari and Opera the last two recent stable releases.

Older versions of the above and other browsers may work, but these are the ones we verify.

Building this repository requires:

* [Node.js](https://nodejs.org/) >=4.0
* [Grunt](https://gruntjs.com/) >=1.0
* [Composer](https://getcomposer.org/)
* cURL and patch CLI programs.

Environment must consist of:

* Apache >=2.2
* PHP >=5.3
* MySQL >=5
* Unix-like OS, e.g. Linux or macOS

## Setup

### Installing required tools

The project uses [Grunt](https://gruntjs.com/) to run tasks and [Sass](http://sass-lang.com/) for CSS pre-processing. This creates a few dependencies in addition to your normal PHP required by FluxBB. First make sure you have base dependencies installed: [Node.js](https://nodejs.org/) and [Grunt](https://gruntjs.com/). You can install Node using the [installer](https://nodejs.org/), Composer using the [installer](https://getcomposer.org/), and Grunt with npm:

```ShellSession
$ npm install -g grunt-cli
```

Consult the Grunt documentation for more instructions if necessary. You might need to use `sudo npm install -g grunt-cli` instead when installing on certain Unix-based systems. Outside from these, the project will also require the normal git, Apache, MySQL and PHP. All requirements must be added to your path variable.

### Setting up Apache virtual host

The `public/` directory is intended to be set as the server's document root. You can do this by adding a new virtual host to your `httpd.conf`. Along the lines of:

```apache
<VirtualHost *:80>
    VirtualHost "/absolute/path/to/textpattern-forum/public"
    ServerName forum.textpattern.test
</VirtualHost>
```

On a local development server, after this you can use your hosts file to point the development domain to correct location (e.g. back to home), run a local DNS server that resolves .test TLDs, or use a [xip.io](http://xip.io/) domain.

### Installing dependencies

After you have the base dependencies taken care of, you can install the project's dependencies. Navigate to the project's directory, and run the dependency managers:

```ShellSession
$ cd textpattern-forum
$ npm install
$ composer install
```

**npm** installs Node modules for Grunt and **composer** installs PHP libraries. You might need to use `sudo` instead when installing on certain Unix-based systems.

### Installing FluxBB

Once you have Grunt installed, installing and updating FluxBB is easy. To setup or update FluxBB run:

```ShellSession
$ grunt setup
```

This will download the correct release version FluxBB, patch it with mods and place it in the `public/` directory. Complete FluxBB's installation by following the normal [Installation steps](http://fluxbb.org/docs/v1.5/installing), and use [this as your config.php template](https://github.com/textpattern/textpattern-forum/blob/master/src/setup/config.php.dist).

After you have finished installing, run the postsetup task to remove extra trash left by the setup:

```ShellSession
$ grunt postsetup
```

### Updating and patching FluxBB

Periodically we update the [patch files](https://github.com/textpattern/textpattern-forum/tree/master/src/setup/patches) that ship with the repository. To re-patch and update your FluxBB installation, run the setup task again:

```ShellSession
$ grunt setup
```

If FluxBB installation version has changed, access your forums and follow the updating steps. After done, run `postsetup` task:

```ShellSession
$ grunt postsetup
```

You may need to [adjust the write permissions](https://fluxbb.org/docs/v1.5/installing#write-permissions) on the `public/cache/` directory after updating.

## Building

This repository hosts sources and needs to be built before it can be used. After you have installed all dependencies, you will be able to run tasks using Grunt, including building:

```ShellSession
$ grunt @task@
```

Where the `@task@` is either `build`, `setup` or `watch`.

* The `build` task builds the project.
* The `setup` task installs the latest stable release of FluxBB (see details above).
* The `watch` task will launch a task that watches for file changes; the project is then automatically built if a source file is modified.

## REST API

We offer a public API that can be used to retrieve data from the forums and topics. The public end point can be found from `https://forum.textpattern.io/api`. Data can be retrieved as JSON, and options can be passed optional HTTP query parameters. This API is read-only, and all requests must be done using HTTP GET method. You can also ping resource existence with HEAD, but any other methods are rejected.

### Response codes

HTTP status codes can be used to detect the type of response. The server responds with:

* 200: Successful response.
* 401: You will need to log in to access this resource. The resource may be removed, or private.
* 404: The resource has been removed, or has never existed.
* 500: Internal server error.
* 503: Server is under maintenance.

### Authentication

Authentication happens using basic HTTP method and regular forum user account.

```ShellSession
$ curl -u "username" https://forum.textpattern.io/api/
```

If you request a private or removed resource without being logged in, you will be treated with 401 response:

```ShellSession
$ curl -I https://forum.textpattern.io/api/posts/401
```

Responds with:

```
HTTP/1.0 401 Unauthorized
Date: Tue, 15 Oct 2013 09:16:42 GMT
Server: Apache
WWW-Authenticate: Basic realm="Textpattern CMS Support Forum External Syndication"
Connection: close
Content-Type: application/json; charset=UTF-8
```

### Topics

```
GET https://forum.textpattern.io/api/topics/:forum
```

Returns topics from the specified forum.

#### Optional parameters

* **limit**: number items to show. A value between 1 and 50. Defaults to 15.
* **sort**: either `last_post` or `posted`.

#### Example request

```ShellSession
$ curl https://forum.textpattern.io/api/topics/2
```

Response headers:

```
HTTP/1.1 200 OK
Date: Sun, 13 Oct 2013 11:24:49 GMT
Server: Apache
Content-Type: application/json; charset=UTF-8
```

Response body:

```
{
  "url": ["https:\/\/forum.textpattern.io\/index.php"],
  "topic": [
    {
      "title": "Better way to upload media",
      "link": "https:\/\/forum.textpattern.io\/viewtopic.php?id=40096&action=new",
      "content": "<p>+ lots<\/p>",
      "author": {
        "name": "tye",
        "uri": "https:\/\/forum.textpattern.io\/profile.php?id=5751"
      },
      "posted": "26 September 2013 21:51",
      "postedutc": "2013-09-24T16:18:48Z",
      "id": "40096"
    }
  ]
}
```

### Posts

```
GET https://forum.textpattern.io/api/posts/:topic
```

Returns replies from a specific public topic.

#### Optional parameters

* **limit**: number items to show. A value between 1 and 50. Defaults to 15.

#### Example request

```ShellSession
$ curl https://forum.textpattern.io/api/posts/40092
```

Response headers:

```
HTTP/1.1 200 OK
Date: Sun, 13 Oct 2013 12:24:49 GMT
Server: Apache
Content-Type: application/json; charset=UTF-8
```

Response body:

```
{
  "url": ["https:\/\/forum.textpattern.io\/viewtopic.php?id=40092"],
  "post": [
    {
      "title": "Re: Tag tree best practise: to nest or not to nest",
      "link": "https:\/\/forum.textpattern.io\/viewtopic.php?pid=275473#p275473",
      "content": "<p>Stef, this is really useful &#8211; I&#8217;m on the right track. Thank you very much.<\/p>",
      "author": {
        "name": "gaekwad",
        "uri": "https:\/\/forum.textpattern.io\/profile.php?id=7456"
      },
      "posted": "26 September 2013 21:51",
      "postedutc": "2013-09-24T16:18:48Z",
      "id": "275473"
    }
  ]
}
```

## Contributing

Want to help out with the development of Textpattern CMS? Please refer to the [Contributing documentation](https://docs.textpattern.io/development/contributing) for full details.

## License

Licensed under MIT license.


CHANGELOG.md
# Changelog

Here's a summary of changes in each update.

## Version 1.1.4 - UPCOMING

* Fixed: Link underline styling.
* Changed: Use the official FluxBB reCAPTCHA plugin, instead of bespoke reCAPTCHA code.
* Changed: Navigation focus style improvements.
* Changed: Better tabs layout.
* Changed: Consistent pagination style with upcoming main site.
* Changed: Fonts served from local server, not Google Web Fonts. Also, use `latin-extended` version of PT Serif for wider language support.
* Changed: Forum is now `utf8mb4` - allows emojis.
* Changed: Style the `<noscript>` notice.
* Changed: Content Security Policy `frame-src` deprecated, now using `child-src`.
* Changed: Support Node v8 and NPM v5.
* Removed: `role="contentinfo"` not required on `<footer>` (source: W3C Validator).
* Removed: DMOZ `noodp` and Yahoo Directory `noydir` are dead.
* Removed: PHP_CodeSniffer - not using it in dev.

## Version 1.1.3 - 2017/03/27

* Fixed: Scrollbar colour on iOS/MacOS.
* Fixed: Encode JSON-LD properly.
* Changed: Larger post edit textarea.
* Changed: Using Greenkeeper to track dependency updates.
* Changed: Customised the XML output slightly, for our Textpattern homepage work.
* Changed: Use the official jQuery CDN.
* Removed: Facebook channel.

## Version 1.1.2 - 2017/03/07

* Changed: Forum moved to new domain (forum.textpattern.io) and now uses HTTPS protocol.
* Changed: Updated Textile to v3.6.
* Added: Use Header Strict Transport Security (HSTS).

## Version 1.1.1 - 2017/02/10

* Fixed: fragmentation of Twitter Card/Open Graph meta.
* Changed: Updated to jQuery v3.x.
* Changed: Code highlighter switched from Google Code Prettify to Prism. Added some more languages support and removed some that are not relevant to our forums.
* Removed: rules page from sitemap (as it's not indexed by search engines).

## Version 1.1.0 - 2017/02/07

* Fixed: CSP blocking base64 encoded icon.
* Changed: Truncate long code boxes.
* Changed: Migrated to Google Universal Analytics.
* Changed: Various performance optimisations and improvements.
* Changed: Improvements to theme style, in line with work on other official sites development.
* Changed: Removed news scraper site, added Awesome Textpattern link in footer.
* Added: Twitter cards.
* Added: Description meta throughout.
* Added: JSON-LD.
* Added: Open Graph meta.
* Added: Robots meta 
* Added: Basic sitemap.xml.
* Added: forum Twitter link.
* Removed: autosize.
* Removed: Discretionary ligatures.
* Removed: Ligatures on monospace font.

## Version 1.0.0 - 2016/11/04

* Fixed: FluxBB v1.5.10. Bug fixes and security improvements.
* Changed: Updated style for consistency with other official sites development.
* Changed: New URL for 'documentation'.
* Changed: New URL for 'themes'.
* Changed: Completely overhauled the build process.
* Changed: Removed Ruby, Compass, Toolkit, Singularity.gs and Sass-Breakpoint from build.
* Removed: Meta data for favicons/apple-touch-icons/Windows tiles that are no longer required.

## Version 0.3.6 - 2015/10/12

* Fixed: Search form should have an ARIA role.
* Fixed: accessibilityHazard spec has changed.
* Changed: Dropped support for Internet Explorer 8.
* Changed: TXP Tips is now Textpattern Tips.
* Changed: minor CSS code optimisations.

## Version 0.3.5 - 2015/04/11

* Fixed: Search microdata errors.
* Changed: New URL for where the logo will be in new .com site.

## Version 0.3.4 - 2015/04/11

* Changed: Improvements to ARIA and microdata.
* Changed: Updated high resolution media query syntax.
* Changed: Updated dependencies.
* Added: editorconfig file.
* Added: More comments in build files.

## Version 0.3.3 - 2015/04/07

* Changed: Updated dependencies.

## Version 0.3.2 - 2015/01/29

* Fixed: FluxBB v1.5.8. Fixes bugs, adds an `addons` directory for rudimentary plugin support, plus other minor improvements.
* Added: Extended microdata markup of HTML structure.
* Added: Meta `theme-color` tag.
* Changed: Performance improvements.
* Changed: Serve HTML5Shiv from CDNJS.
* Changed: Updated jQuery versions.
* Changed: Updated various build components.

## Version 0.3.1 - 2014/11/10

* Fixed: Content-Security-Policy headers were failing to load Google Web Fonts, because Google have recently changed the font serving URL.

## Version 0.3.0 - 2014/11/07

* Fixed: FluxBB v1.5.7. Fixes security vulnerabilities.
* Fixed: Double-border glitch in IE 11.
* Fixed: Prevent really long words from breaking alert box layouts.
* Fixed: Prevent IE 8 layout from collapsing too far.
* Fixed: Textile v3.5.5.
* Added: Theme source code has been extensive commented throughout.
* Added: New apple touch icon size for iOS 8.
* Changed: Speed up CSS transitions.
* Changed: Remove legacy touch icons.
* Changed: Updated various components and Grunt tasks.
* Changed: Slightly more room by removing left-padding of tables (helps reduce likelihood of layout spillage on smallest devices).
* Changed: Various minor CSS improvements and optimisations.

## Version 0.2.0 - 2014/01/11

* Fixed: FluxBB v1.5.6. Fixes account renaming, HTML5 validation and security vulnerabilities.
* Fixed: Masthead logo glitch on IE 8. Doesn't display Textpattern logo twice.
* Fixed: Main network navigation loading and content shifting.
* Fixed: Improve font loading on subsequent page visits. Move from Web Font Loader to a link tag.
* Fixed: Remove JavaScript based moving of the subscribe button; causes content shifting.
* Fixed: Off-positioned forum navigation tabs. Appeared 1px below the strip border on Chrome.
* Fixed: Don't send non-IPv4 addresses to Stop Forum Spam.
* Fixed: Sfs::isBanned() check; do not create duplicate ban records.
* Fixed: Only load HTML5 Shiv on IE 8.
* Fixed: Rare cases where there is no space rendered between forum navigation tabs.
* Fixed: Issues where wrong syntax highlighting language would be used.
* Fixed: Fix code block quoting and its escaping.

src/style/Textpattern/js/main.js
(function ()
{
    'use strict';

    document.documentElement.className = 'js';

    requirejs.config({
        paths: {
            'jquery': 'https://code.jquery.com/jquery-3.2.1.min'
        }
    });

    // Detect whether user enabled 'Do No Track' in their browser, and honour it.

    define('track', function ()
    {
        return {
            allow : navigator.doNotTrack !== 'yes' && navigator.doNotTrack !== '1' && window.doNotTrack !== 'yes' && window.doNotTrack !== '1'
        };
    });

    // Quoting.

    require(['jquery'], function ($)
    {
        var field = $('#quickpostform textarea[name=req_message]'), button;

        if (!field.length) {
            return;
        }

        $('.postlink a').eq(0).attr('href', '#quickpostform');

        button = $('<a href="#quickpostform">Quote</a>').on('click', function ()
        {
            var $this = $(this),
                tagStart,
                tagEnd,
                paragraph,
                post = $this.parents('.blockpost').eq(0),
                name = post.find('.postleft dl dt').eq(0).text(),
                message = post.find('.postmsg').eq(0).clone().find('.postedit, table').remove().end(),
                link = post.find('h2 a').eq(0).attr('href'),
                postId = post.attr('id').substr(1),
                value = $.trim(field.val());

            // Remove quotes.

            message.find('blockquote').each(function ()
            {
                var bq = $(this), prev = bq.prev('h6, p');

                if (prev.length && $.trim(prev.text()).substr(-1) === ':') {
                    prev.remove();
                }

                bq.remove();
            });

            // Compress code blocks to a single line.

            message.find('pre').each(function ()
            {
                $(this).after('<p>@' + $.trim($('<div />').text($(this).text()).html().split('\n').slice(0, 1).join('')) + '...@</p>').remove();
            });

            // Links.

            message.find('a').each(function ()
            {
                $(this).prepend('"').append('":' + $('<div />').text($(this).attr('href')).html());
            });

            // Images.

            message.find('img').each(function ()
            {
                $(this).after('!' + $('<div />').text($(this).attr('src')).html() + '!').remove();
            });

            // Spans.

            message.find('strong').prepend('*').append('*');
            message.find('b').prepend('**').append('**');
            message.find('cite').prepend('??').append('??');
            message.find('em').prepend('_').append('_');
            message.find('i').prepend('__').append('__');
            message.find('del').prepend('-').append('-');
            message.find('ins').prepend('+').append('+');
            message.find('sub').prepend('~').append('~');
            message.find('sup').prepend('^').append('^');
            message.find('code').prepend('@').append('@');

            // Headings.

            message.find('h1, h2, h3, h4, h5, h6').prepend('*').append('*');

            // Lists. Textile will wrap the lists in paragraphs as of 3.5.x.

            $.each({'ol' : '#', 'ul' : '*'}, function (type, marker)
            {
                message.children(type).each(function ()
                {
                    var items = [], list = $(this);

                    list.find('li').each(function ()
                    {
                        var bullet = '';

                        for (var i = 0; i < $(this).parents(list).length / 2; i++) {
                            bullet += marker;
                        }

                        items.push(bullet + ' ' + $.trim($(this).clone().children(type).remove().end().html()));
                    });

                    list.after('<p>' + items.join('\n') + '</p>').remove();
                });
            });

            paragraph = message.find('h1, h2, h3, h4, h5, h6, p');

            if (paragraph.length > 1) {
                tagStart = '\n\nbq.. ';
                tagEnd = '\n\np. ';
            } else {
                tagStart = '\n\nbq. ';
                tagEnd = '\n\n';
            }

            message = $.trim(paragraph.append('\n\n').text());

            if (value) {
                value = $.trim(value).replace(/\r?\n\r?\np\.$/, '') + '\n\n';
            }

            name = $.trim(name).replace(/==/, '');

            if (message) {
                value = value + 'h6. ==' + name + '== wrote "#' + postId + '":./' + link + ':' + tagStart + message + tagEnd;
            } else {
                value = value + 'h6. In reply to ==' + name + '== "#' + postId + '":./' + link + ':\n\n';
            }

            field.val(value).trigger('autosize.resize').focus();

            if ($.type(field[0].setSelectionRange) !== 'undefined') {
                field[0].setSelectionRange(value.length, value.length);
            }
        });

        $('.postfootright ul').append($('<li class="textile-quote-post" />').html($('<span />').html(button)));
    });

    // Ranks.

    require(['jquery'], function ($)
    {
        var ranks =
        [
            {
                posts : 0,
                title : 'alpha'
            },
            {
                posts : 10,
                title : 'beta'
            },
            {
                posts : 35,
                title : 'gamma'
            },
            {
                posts : 50,
                title : 'delta'
            },
            {
                posts : 75,
                title : 'epsilon'
            },
            {
                posts : 100,
                title : 'zêta'
            },
            {
                posts : 150,
                title : 'êta'
            },
            {
                posts : 250,
                title : 'thêta'
            },
            {
                posts : 300,
                title : 'iota'
            },
            {
                posts : 350,
                title : 'kappa'
            },
            {
                posts : 400,
                title : 'lambda'
            },
            {
                posts : 500,
                title : 'mu'
            },
            {
                posts : 600,
                title : 'nu'
            },
            {
                posts : 700,
                title : 'xi'
            },
            {
                posts : 800,
                title : 'omicron'
            },
            {
                posts : 900,
                title : 'pi'
            },
            {
                posts : 1000,
                title : 'rho'
            },
            {
                posts : 1100,
                title : 'sigma'
            },
            {
                posts : 1200,
                title : 'tau'
            },
            {
                posts : 1300,
                title : 'upsilon'
            },
            {
                posts : 1400,
                title : 'phi'
            },
            {
                posts : 1500,
                title : 'chi'
            },
            {
                posts : 1600,
                title : 'psi'
            },
            {
                posts : 1700,
                title : 'omega'
            },
            {
                posts : 2000,
                title : 'zero'
            },
            {
                posts : 2100,
                title : 'ichi'
            },
            {
                posts : 2200,
                title : 'ni'
            },
            {
                posts : 2300,
                title : 'san'
            },
            {
                posts : 2400,
                title : 'yon'
            },
            {
                posts : 2500,
                title : 'go'
            },
            {
                posts : 2600,
                title : 'roku'
            },
            {
                posts : 2700,
                title : 'nana'
            },
            {
                posts : 2800,
                title : 'hachi'
            },
            {
                posts : 2900,
                title : 'kyū'
            },
            {
                posts : 3000,
                title : 'Mau'
            },
            {
                posts : 3005,
                title : 'Octopus'
            },
            {
                posts : 3006,
                title : '0ctocat'
            },
            {
                posts : 3077,
                title : 'Mautwo'
            },
            {
                posts : 3100,
                title : 'Byakko'
            },
            {
                posts : 3200,
                title : 'Kasha'
            },
            {
                posts : 3300,
                title : 'Bakeneko'
            },
            {
                posts : 3400,
                title : 'Nue'
            },
            {
                posts : 3500,
                title : 'Shachihoko'
            },
            {
                posts : 4000,
                title : 'Isuteritosu'
            },
            {
                posts : 4200,
                title : 'Ogopogo'
            },
            {
                posts : 4300,
                title : 'Wyvern'
            },
            {
                posts : 4400,
                title : 'Serpion'
            },
            {
                posts : 4444,
                title : 'Kusanagi'
            },
            {
                posts : 4600,
                title : 'Leviathan'
            },
            {
                posts : 4700,
                title : 'Zalamander'
            },
            {
                posts : 4800,
                title : 'Gizamaluk'
            },
            {
                posts : 4900,
                title : 'Vyraal'
            },
            {
                posts : 5000,
                title : 'Abelisk'
            },
            {
                posts : 5555,
                title : 'Sugari no Ontachi'
            },
            {
                posts : 5600,
                title : 'Tiamat'
            },
            {
                posts : 5800,
                title : 'Shinryu'
            },
            {
                posts : 6000,
                title : 'Bahamut'
            },
            {
                posts : 6500,
                title : 'Kirin'
            },
            {
                posts : 6666,
                title : 'Masamune'
            },
            {
                posts : 6800,
                title : 'Ryū'
            },
            {
                posts : 7000,
                title : 'Chrysophylax'
            },
            {
                posts : 7500,
                title : 'Glaurung'
            },
            {
                posts : 8000,
                title : 'Icefyre'
            },
            {
                posts : 8500,
                title : 'Mizuchi'
            },
            {
                posts : 9000,
                title : 'Kuzuryū'
            },
            {
                posts : 9500,
                title : 'Azure'
            },
            {
                posts : 10000,
                title : 'Mizuchi'
            },
            {
                posts : 10500,
                title : 'Watatsumi'
            },
            {
                posts : 11000,
                title : 'Ryūjin'
            },
            {
                posts : 11500,
                title : 'Zennyo Ryūō'
            },
            {
                posts : 12000,
                title : 'Amaterasu'
            },
            {
                posts : 30246,
                title : 'Code Is Pottery'
            },
            {
                posts : 36625,
                title : 'Sandal I\'d Own'
            }
        ];

        $('#page-viewtopic .blockpost .postleft > dl').each(function ()
        {
            var i, rank, $this = $(this), posts = parseInt($this.find('dd span').filter(function ()
            {
                return $(this).text().indexOf('Posts: ') === 0;
            }).eq(0).text().replace('Posts: ', '').replace(',', ''), 10);

            for (i = 0; i < ranks.length; i++) {
                if (ranks[i].posts > posts) {
                    break;
                }

                rank = ranks[i];
            }

            if (rank) {
                $this.find('.usertitle').after('<dd class="userrank"><span title="Ranked level '+ (i - 1) +'">'+rank.title+'</span></dd>');
            }
        });
    });

    // Syntax highlighting, via 'Prism'.
    // Applies syntax highlighting to `code` HTML elements.
    // More info - http://prismjs.com.

    require(['jquery'], function ($)
    {
        var code = $('.prism');

        if (code.length) {
            code.filter('.language-txp').addClass('language-html').removeClass('language-txp');

            require(['prism.@@timestamp'], function ()
            {
                Prism.highlightAll();
            });
        }
    });

    // Focus to the username/password on the login page.

    require(['jquery'], function ($)
    {
        if (!$('#page-login').length) {
            return;
        }

        var user = $('#login [name=req_username]'), pass = $('#login [name=req_password]');

        if (user.val() === '' || pass.val() !== '') {
            user.focus();
        } else {
            pass.focus();
        }
    });

    // Forms.

    require(['jquery'], function ($)
    {
        $('input[type!=hidden][name^=req_], textarea[name^=req_]').prop('required', true);

        $('#qjump select').change(function ()
        {
            window.location = 'viewforum.php?id=' + $(this).val();
        });
    });

    // Responsive navigation menu, via 'Responsive Nav'.
    // More info - https://github.com/viljamis/responsive-nav.js.

    require(['responsivenav.@@timestamp'], function ()
    {
        responsiveNav('.site-navigation', {
            transition: 400,
            insert: 'before',
            navClass: 'site-navigation'
        });
    });

    // Embed widgets.

    require(['jquery'], function ($)
    {
        var gistStyle = false,
            loadTwitter = false,
            tweetRegex = /^https?:\/\/twitter\.com\/(#!\/)?[a-z0-9]+\/status(es)?\/[0-9]+$/i,
            gistRegex = /^https?:\/\/gist\.github\.com\/[a-z0-9]+\/[0-9]+$/i,
            youtubeRegex = /^https?:\/\/(?:www\.)?(?:youtube\.com\/watch(?:\/|\?v=)|youtu\.be\/)([a-z0-9\-\_]+)$/i;

        $('.postmsg > p > a').each(function ()
        {
            var $this = $(this), href = $this.attr('href'), matches;

            if ($this.parent().text() !== $this.text()) {
                return;
            }

            if (gistRegex.test(href)) {
                $.ajax(href + '.json', {dataType: 'jsonp'})
                    .done(function (data)
                    {
                        if (data && data.div) {
                            if (gistStyle === false && data.stylesheet) {
                                $('head').append($('<link rel="stylesheet" />').attr('href', 'https://gist.github.com' + data.stylesheet));
                                gistStyle = true;
                            }

                            $this.parent().after($(data.div).removeAttr('id')).hide();
                        }
                    });

                return;
            }

            if (tweetRegex.test(href)) {
                loadTwitter = true;

                $this.parent().after(
                    $('<blockquote class="twitter-tweet" data-dnt="true" />').html($this.parent().html())
                ).hide();

                return;
            }

            matches = href.match(youtubeRegex);

            if (matches) {
                $this.parent().after(
                    $('<div class="embed-video embed-youtube" />').html(
                        $('<iframe frameborder="0" allowfullscreen></iframe>').attr('src', 'https://www.youtube-nocookie.com/embed/' + matches[1])
                    )
                ).hide();

                return;
            }
        });

        if (loadTwitter) {
            $('head').append('<meta name="twitter:widgets:csp" content="on" />');
            require(['https://platform.twitter.com/widgets.js']);
        }
    });

    // Google Analytics

    require(['track'], function(track)
    {
        if (track.allow) {
            /* jshint ignore:start */
            (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
            (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
            })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
            /* jshint ignore:end */
            ga('create', 'UA-191562-28', 'auto', {
                anonymizeIp: true
            });
            ga('send', 'pageview');
        }
    });

})();
